src/stats.rs

```diff
@@ -1,45 +1,219 @@
-static LATEST_STATS: Lazy<Mutex<HashMap<usize, HashMap<String, i64>>>> =
-    Lazy::new(|| Mutex::new(HashMap::new()));
-
 /// Statistics period used for average calculations.
 /// 15 seconds.
 static STAT_PERIOD: u64 = 15000;
 
+/// The various states that a client can be in
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum ClientState {
+    Idle,
+    Waiting,
+    Active,
+}
+impl std::fmt::Display for ClientState {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            ClientState::Idle => write!(f, "idle"),
+            ClientState::Waiting => write!(f, "waiting"),
+            ClientState::Active => write!(f, "active"),
+        }
+    }
+}
+
+/// The various states that a server can be in
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum ServerState {
+    Login,
+    Active,
+    Tested,
+    Idle,
+}
+impl std::fmt::Display for ServerState {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            ServerState::Login => write!(f, "login"),
+            ServerState::Active => write!(f, "active"),
+            ServerState::Tested => write!(f, "tested"),
+            ServerState::Idle => write!(f, "idle"),
+        }
+    }
+}
+
+/// Information we keep track off which can be queried by SHOW CLIENTS
+#[derive(Debug, Clone)]
+pub struct ClientInformation {
+    pub state: ClientState,
+    pub connect_time: Instant,
+
+    /// A random integer assigned to the client and used by stats to track the client
+    pub client_id: i32,
+
+    pub application_name: String,
+    pub username: String,
+    pub pool_name: String,
+
+    /// Total time spent waiting for a connection from pool, measures in microseconds
+    pub total_wait_time: u64,
+
+    pub transaction_count: u64,
+    pub query_count: u64,
+    pub error_count: u64,
+}
+
+/// Information we keep track off which can be queried by SHOW SERVERS
+#[derive(Debug, Clone)]
+pub struct ServerInformation {
+    pub state: ServerState,
+    pub connect_time: Instant,
+
+    /// A random integer assigned to the server and used by stats to track the server
+    pub server_id: i32,
+
+    pub address_name: String,
+    pub address_id: usize,
+
+    pub username: String,
+    pub pool_name: String,
+    pub application_name: String,
+
+    pub bytes_sent: u64,
+    pub bytes_received: u64,
+
+    pub transaction_count: u64,
+    pub query_count: u64,
+    pub error_count: u64,
+}
+
 /// The names for the events reported
 /// to the statistics collector.
-#[derive(Debug, Clone, Copy)]
+#[derive(Debug, Clone)]
 enum EventName {
-    CheckoutTime,
-    Query,
-    Transaction,
-    DataSent,
-    DataReceived,
-    ClientWaiting,
-    ClientActive,
-    ClientIdle,
-    ClientDisconnecting,
-    ServerActive,
-    ServerIdle,
-    ServerTested,
-    ServerLogin,
-    ServerDisconnecting,
-    UpdateStats,
-    UpdateAverages,
+    CheckoutTime {
+        client_id: i32,
```

Comment:
instead of adding only the important values to this struct could we instead use a more general struct like client_metadata which holds all the info we might need about a client and probably just use address for server info. This way if we ever want to add tagging to statsd we can do so really easily

Reply:
That would require constructing the client metadata object on every call site. Sometimes all we have is just `client_id`

Reply:
This is what I used in my here, https://github.com/levkk/pgcat/blob/1d32eefe633572f25165670b1c84d56a486f2f2e/src/stats.rs#L48

Reply:
The `client_register` event is the only place where we need to add meta data. The meta data is stored locally and the collector and we do a lookup on `client_id` every time an event comes

Reply:
The way I did it was just creating it once and then cloning it when needed (clone required cause of strings)

Reply:
The equivalent of the client_metadata here would be the `ClientInformation` struct that we populated at client_register event

Reply:
Client and server information objects
https://github.com/levkk/pgcat/blob/08af06d6402f31fb105ea01bf656667c7eff7099/src/stats.rs#L85-L128